From 025182f3b46cfb62cc2f9b742fbf9d7ef7508e84 Mon Sep 17 00:00:00 2001
From: Jing Hua <jinghua@marvell.com>
Date: Fri, 13 Jun 2014 10:17:50 +0800
Subject: [PATCH 1719/1825] fix: alp: tpm: when merge several CLS hit into one
 C2 lu_type, mixed the vlan num

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 01df8ca811e81e48e371a20d7d69f4c444e96c83

	In routine tpm_mng_cap_merge_flow, when merge several CAPs into one C2
	lu_type, mixed the vlan num by mistake. So HWF rules for single tag and
	double tag use the same lu_type. To fix this, do not merge CAPs with
	different VLAN number
	SYSTEMSW-658 - <TPM cant support vid value and vlan num in the same hardware forward rule>

Signed-off-by: Jing Hua <jinghua@marvell.com>

Change-Id: I2ee53e557770a697b70e65e7d20fcbaf52515442
Reviewed-on: http://vgitil04.il.marvell.com:8080/8522
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Dmitri Epshtein <dima@marvell.com>
Reviewed-by: Evan Wang <xswang@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 .../mv_drivers_lsp/mv_tpm/src/conf/tpm_mng.c       |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/arch/arm/plat-armada/mv_drivers_lsp/mv_tpm/src/conf/tpm_mng.c b/arch/arm/plat-armada/mv_drivers_lsp/mv_tpm/src/conf/tpm_mng.c
index 0832bdb..ad35f79 100644
--- a/arch/arm/plat-armada/mv_drivers_lsp/mv_tpm/src/conf/tpm_mng.c
+++ b/arch/arm/plat-armada/mv_drivers_lsp/mv_tpm/src/conf/tpm_mng.c
@@ -2162,6 +2162,7 @@ int tpm_mng_cap_merge_flow(
 	int cls_pri;
 	int field_bm;
 	enum tpm_engine_no_t eng_no;
+	enum tpm_vlan_num_enum_t vlan_num;
 
 	IF_NULL(TPM_MNG_MOD, cap_arr_in);
 	IF_NULL(TPM_MNG_MOD, cap_arr_out);
@@ -2187,10 +2188,12 @@ int tpm_mng_cap_merge_flow(
 		field_bm = cap_arr_in[loop].cap_field_bm;
 		c2_pri = cap_arr_in[loop].c2_pri;
 		cls_pri = cap_arr_in[loop].cls_pri;
+		vlan_num = cap_arr_in[loop].vlan_num;
 		while (loop_in < cap_num) {
 			/* if it is not C2 cap, or is not CLS cap, or does not have field_bm, could not be merged */
 			if ((cap_arr_in[loop_in].eng_no != TPM_ENGINE_C2)
-			     || (cap_arr_in[loop].cap_api_type != TPM_API_PUBLIC_CLS)
+			     || (cap_arr_in[loop_in].cap_api_type != TPM_API_PUBLIC_CLS)
+			     || (cap_arr_in[loop_in].vlan_num != vlan_num)
 			     || (cap_arr_in[loop_in].cap_field_bm == 0))
 				break;
 
-- 
1.7.5.4

